File: python/src/evalis/ast.py
from evalis.__gen__.EvalisVisitor import EvalisVisitor as BaseEvalisVisitor
from evalis.__gen__.EvalisParser import EvalisParser
from .types import (
    BinaryOpNode,
    BinaryOpType,
    UnaryOpNode,
    UnaryOpType,
    LiteralNode,
    ReferenceNode,
    EvalisNode,
)


def _get_op_text(ctx):
    """
    Helper to return ctx.op.text, because this wasn't passing the type checker
    """
    return ctx.op.text


class AstBuilder(BaseEvalisVisitor):
    def visitParse(self, ctx):
        return self.visit(ctx.expr())

    # Visit a parse tree produced by EvalisParser#AndExpr.
    def visitAndExpr(self, ctx):
        return BinaryOpNode(
            op=BinaryOpType.AND,
            left=self.visit(ctx.expr(0)),
            right=self.visit(ctx.expr(1)),
        )

    # Visit a parse tree produced by EvalisParser#MulDivExpr.
    def visitMulDivExpr(self, ctx: EvalisParser.MulDivExprContext):
        return BinaryOpNode(
            op=BinaryOpType(_get_op_text(ctx)),
            left=self.visit(ctx.expr(0)),
            right=self.visit(ctx.expr(1)),
        )

    # Visit a parse tree produced by EvalisParser#EqualityExpr.
    def visitEqualityExpr(self, ctx: EvalisParser.EqualityExprContext):
        return BinaryOpNode(
            op=BinaryOpType(_get_op_text(ctx)),
            left=self.visit(ctx.expr(0)),
            right=self.visit(ctx.expr(1)),
        )

    # Visit a parse tree produced by EvalisParser#NotExpr.
    def visitNotExpr(self, ctx: EvalisParser.NotExprContext):
        return UnaryOpNode(op=UnaryOpType.NOT, expr=self.visit(ctx.expr()))

    # Visit a parse tree produced by EvalisParser#RelationalExpr.
    def visitRelationalExpr(self, ctx: EvalisParser.RelationalExprContext):
        return BinaryOpNode(
            op=BinaryOpType(_get_op_text(ctx)),
            left=self.visit(ctx.expr(0)),
            right=self.visit(ctx.expr(1)),
        )

    def visitAtomExpr(self, ctx: EvalisParser.AtomExprContext):
        # LiteralNode
        if ctx.atom().literal():
            return self.visit(ctx.atom().literal())

        # Parentheses
        if ctx.atom().expr():
            return self.visit(ctx.atom().expr())

        # Identifier (+ optional access suffixes)
        base_identifier = ctx.atom().identifier().getText()
        parts: list[EvalisNode] = []

        for suffix in ctx.atom().accessSuffix():
            # Dot access
            if suffix.identifier():
                parts.append(LiteralNode(suffix.identifier().getText()))
            # Index access (e.g., ["key"])
            elif suffix.expr():
                parts.append(self.visit(suffix.expr()))

        return ReferenceNode(root=base_identifier, children=tuple(parts))

    # Visit a parse tree produced by EvalisParser#AddSubExpr.
    def visitAddSubExpr(self, ctx: EvalisParser.AddSubExprContext):
        return BinaryOpNode(
            op=BinaryOpType(_get_op_text(ctx)),
            left=self.visit(ctx.expr(0)),
            right=self.visit(ctx.expr(1)),
        )

    # Visit a parse tree produced by EvalisParser#OrExpr.
    def visitOrExpr(self, ctx: EvalisParser.OrExprContext):
        return BinaryOpNode(
            op=BinaryOpType.OR,
            left=self.visit(ctx.expr(0)),
            right=self.visit(ctx.expr(1)),
        )

    def visitInExpr(self, ctx: EvalisParser.InExprContext):
        return BinaryOpNode(
            op=BinaryOpType.IN,
            left=self.visit(ctx.expr(0)),
            right=self.visit(ctx.expr(1)),
        )

    # Visit a parse tree produced by EvalisParser#number.
    def visitNumber(self, ctx: EvalisParser.NumberContext):
        text: str = ctx.getText()

        if "." in text:
            return LiteralNode(float(text))

        return LiteralNode(int(text))

    # Visit a parse tree produced by EvalisParser#boolean.
    def visitBoolean(self, ctx: EvalisParser.BooleanContext):
        return LiteralNode(ctx.getText() == "true")

    # Visit a parse tree produced by EvalisParser#stringLiteralNode.
    def visitStringLiteral(self, ctx: EvalisParser.StringLiteralContext):
        raw = ctx.getText()
        unquoted = raw[1:-1]
        unescaped = unquoted.replace('\\"', '"').replace("\\\\", "\\")
        return LiteralNode(unescaped)

File: typescript/src/ast.ts
import { ParseTreeVisitor } from 'antlr4';
import * as Parser from './__gen__/EvalisParser';
import EvalisVisitor from './__gen__/EvalisVisitor';
import { BinaryOpType, UnaryOpType } from './__gen__/grammar';
import { EvalisNode } from './types';

export class AstBuilder
  extends ParseTreeVisitor<EvalisNode>
  implements EvalisVisitor<EvalisNode>
{
  protected defaultResult(): EvalisNode {
    throw new Error('No default result');
  }

  visitParse(ctx: Parser.ParseContext) {
    return this.visit(ctx.expr());
  }

  visitAndExpr(ctx: Parser.AndExprContext): EvalisNode {
    return {
      type: 'binaryOp',
      op: BinaryOpType.AND,
      left: this.visit(ctx.expr(0)),
      right: this.visit(ctx.expr(1)),
    };
  }

  visitMulDivExpr(ctx: Parser.MulDivExprContext): EvalisNode {
    const opText = ctx._op?.text;
    if (!opText) throw new Error('Missing operator');

    return {
      type: 'binaryOp',
      op: opText as BinaryOpType,
      left: this.visit(ctx.expr(0)),
      right: this.visit(ctx.expr(1)),
    };
  }

  visitEqualityExpr(ctx: Parser.EqualityExprContext): EvalisNode {
    const opText = ctx._op?.text;
    if (!opText) throw new Error('Missing operator');

    return {
      type: 'binaryOp',
      op: opText as BinaryOpType,
      left: this.visit(ctx.expr(0)),
      right: this.visit(ctx.expr(1)),
    };
  }

  visitNotExpr(ctx: Parser.NotExprContext): EvalisNode {
    return {
      type: 'unaryOp',
      op: UnaryOpType.NOT,
      expr: this.visit(ctx.expr()),
    };
  }

  visitRelationalExpr(ctx: Parser.RelationalExprContext): EvalisNode {
    const opText = ctx._op?.text;
    if (!opText) throw new Error('Missing operator');

    return {
      type: 'binaryOp',
      op: opText as BinaryOpType,
      left: this.visit(ctx.expr(0)),
      right: this.visit(ctx.expr(1)),
    };
  }

  visitAtomExpr(ctx: Parser.AtomExprContext): EvalisNode {
    return this.visit(ctx.atom());
  }

  visitLiteralAtom(ctx: Parser.LiteralAtomContext): EvalisNode {
    return this.visit(ctx.literal());
  }

  visitParenAtom(ctx: Parser.ParenAtomContext): EvalisNode {
    return this.visit(ctx.expr());
  }

  visitIdentifierAtom(ctx: Parser.IdentifierAtomContext): EvalisNode {
    const baseIdentifier = ctx.identifier().getText();
    const parts: EvalisNode[] = [];

    for (const suffix of ctx.accessSuffix_list()) {
      if (suffix.identifier()) {
        parts.push({
          type: 'literal',
          value: suffix.identifier()!.getText(),
        });
      } else if (suffix.expr()) {
        parts.push(this.visit(suffix.expr()!));
      }
    }

    return {
      type: 'reference',
      root: baseIdentifier,
      children: parts,
    };
  }

  visitListComprehension(ctx: Parser.ListComprehensionContext): EvalisNode {
    return {
      type: 'listComprehension',
      elementExpr: this.visit(ctx.expr(0)),
      variableName: ctx.identifier().getText(),
      iterableExpr: this.visit(ctx.expr(1)),
    };
  }

  visitAddSubExpr(ctx: Parser.AddSubExprContext): EvalisNode {
    const opText = ctx._op?.text;
    if (!opText) throw new Error('Missing operator');

    return {
      type: 'binaryOp',
      op: opText as BinaryOpType,
      left: this.visit(ctx.expr(0)),
      right: this.visit(ctx.expr(1)),
    };
  }

  visitOrExpr(ctx: Parser.OrExprContext): EvalisNode {
    return {
      type: 'binaryOp',
      op: BinaryOpType.OR,
      left: this.visit(ctx.expr(0)),
      right: this.visit(ctx.expr(1)),
    };
  }

  visitInExpr(ctx: Parser.InExprContext): EvalisNode {
    return {
      type: 'binaryOp',
      op: BinaryOpType.IN,
      left: this.visit(ctx.expr(0)),
      right: this.visit(ctx.expr(1)),
    };
  }

  visitNumber(ctx: Parser.NumberContext): EvalisNode {
    const text = ctx.getText();

    if (text.includes('.')) {
      return {
        type: 'literal',
        value: parseFloat(text),
      };
    }

    return {
      type: 'literal',
      value: parseInt(text, 10),
    };
  }

  visitBoolean(ctx: Parser.BooleanContext): EvalisNode {
    return {
      type: 'literal',
      value: ctx.getText() === 'true',
    };
  }

  visitStringLiteral(ctx: Parser.StringLiteralContext): EvalisNode {
    const raw = ctx.getText();
    const unquoted = raw.slice(1, -1);
    const unescaped = unquoted.replace(/\\"/g, '"').replace(/\\\\/g, '\\');

    return {
      type: 'literal',
      value: unescaped,
    };
  }

  visitLiteral(ctx: Parser.LiteralContext): EvalisNode {
    if (ctx.number_()) {
      return this.visit(ctx.number_()!);
    }
    if (ctx.stringLiteral()) {
      return this.visit(ctx.stringLiteral()!);
    }
    if (ctx.boolean_()) {
      return this.visit(ctx.boolean_()!);
    }
    // null
    return {
      type: 'literal',
      value: null,
    };
  }
}
